File: backend/app/api/v1/endpoints/notifications.py
from fastapi import APIRouter, Depends, status, HTTPException, Query
from sqlalchemy.orm import Session
import uuid
from typing import Optional

from app.database.session import get_db
from app.schemas.notification import (
    NotificationResponse,
    NotificationListResponse,
    UnreadCountResponse,
    NotificationPreferenceResponse,
    NotificationPreferenceUpdate
)
from app.api.dependencies import get_current_user, get_workspace_member
from app.models.user import User
from app.models.workspace import WorkspaceMember
from app.services.notification import NotificationService

router = APIRouter(tags=["Notifications & Real-Time Delivery"])

@router.get("/notifications", response_model=NotificationListResponse)
def list_notifications(
    status: Optional[str] = Query(None),
    type: Optional[str] = Query(None),
    page: int = Query(1, ge=1),
    page_size: int = Query(50, ge=1, le=100),
    current_user: User = Depends(get_current_user),
    workspace_member: WorkspaceMember = Depends(get_workspace_member),
    db: Session = Depends(get_db)
):
    service = NotificationService(db)
    return service.list_notifications(
        workspace_id=workspace_member.workspace_id,
        user_id=current_user.id,
        status=status,
        type=type,
        page=page,
        page_size=page_size
    )

@router.get("/notifications/unread-count", response_model=UnreadCountResponse)
def get_unread_count(
    current_user: User = Depends(get_current_user),
    workspace_member: WorkspaceMember = Depends(get_workspace_member),
    db: Session = Depends(get_db)
):
    service = NotificationService(db)
    count = service.get_unread_count(
        workspace_id=workspace_member.workspace_id,
        user_id=current_user.id
    )
    return UnreadCountResponse(unread_count=count)

@router.post("/notifications/{notification_id}/read", response_model=NotificationResponse)
def mark_notification_read(
    notification_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    workspace_member: WorkspaceMember = Depends(get_workspace_member),
    db: Session = Depends(get_db)
):
    service = NotificationService(db)
    return service.mark_as_read(
        workspace_id=workspace_member.workspace_id,
        user_id=current_user.id,
        notification_id=notification_id
    )

@router.post("/notifications/read-all", response_model=UnreadCountResponse)
def mark_all_notifications_read(
    current_user: User = Depends(get_current_user),
    workspace_member: WorkspaceMember = Depends(get_workspace_member),
    db: Session = Depends(get_db)
):
    service = NotificationService(db)
    service.mark_all_as_read(
        workspace_id=workspace_member.workspace_id,
        user_id=current_user.id
    )
    return UnreadCountResponse(unread_count=0)

@router.get("/notifications/preferences", response_model=NotificationPreferenceResponse)
def get_notification_preferences(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    service = NotificationService(db)
    return service.get_preferences(user_id=current_user.id)

@router.put("/notifications/preferences", response_model=NotificationPreferenceResponse)
def update_notification_preferences(
    payload: NotificationPreferenceUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    service = NotificationService(db)
    service.update_preference(
        user_id=current_user.id,
        notification_type=payload.notification_type,
        in_app_enabled=payload.in_app_enabled,
        email_enabled=payload.email_enabled,
        push_enabled=payload.push_enabled
    )
    return service.get_preferences(user_id=current_user.id)

File: frontend/src/api/notifications.ts
import { request } from './client';

export interface NotificationItem {
  id: string;
  workspace_id: string;
  recipient_user_id: string;
  actor_user_id?: string | null;
  actor_name?: string | null;
  type: string;
  title: string;
  body: string;
  resource_type?: string | null;
  resource_id?: string | null;
  project_id?: string | null;
  team_id?: string | null;
  comment_id?: string | null;
  status: 'unread' | 'read';
  read_at?: string | null;
  created_at: string;
}

export interface NotificationListResponse {
  total: number;
  unread_count: number;
  page: number;
  page_size: number;
  notifications: NotificationItem[];
}

export interface UnreadCountResponse {
  unread_count: number;
}

export interface NotificationPreferenceItem {
  notification_type: string;
  in_app_enabled: boolean;
  email_enabled: boolean;
  push_enabled: boolean;
}

export interface NotificationPreferenceResponse {
  user_id: string;
  preferences: NotificationPreferenceItem[];
}

export async function getNotifications(params?: {
  status?: string;
  type?: string;
  page?: number;
  page_size?: number;
}): Promise<NotificationListResponse> {
  const query = new URLSearchParams();
  if (params?.status) query.append('status', params.status);
  if (params?.type) query.append('type', params.type);
  if (params?.page) query.append('page', params.page.toString());
  if (params?.page_size) query.append('page_size', params.page_size.toString());

  const qStr = query.toString() ? `?${query.toString()}` : '';
  return request<NotificationListResponse>(`/notifications${qStr}`);
}

export async function getUnreadNotificationCount(): Promise<UnreadCountResponse> {
  return request<UnreadCountResponse>('/notifications/unread-count');
}

export async function markNotificationRead(notificationId: string): Promise<NotificationItem> {
  return request<NotificationItem>(`/notifications/${encodeURIComponent(notificationId)}/read`, {
    method: 'POST',
  });
}

export async function markAllNotificationsRead(): Promise<UnreadCountResponse> {
  return request<UnreadCountResponse>('/notifications/read-all', {
    method: 'POST',
  });
}

export async function getNotificationPreferences(): Promise<NotificationPreferenceResponse> {
  return request<NotificationPreferenceResponse>('/notifications/preferences');
}

export async function updateNotificationPreference(data: {
  notification_type: string;
  in_app_enabled?: boolean;
  email_enabled?: boolean;
  push_enabled?: boolean;
}): Promise<NotificationPreferenceResponse> {
  return request<NotificationPreferenceResponse>('/notifications/preferences', {
    method: 'PUT',
    body: JSON.stringify(data),
  });
}
